### PR TITLE
Events

### PR DESCRIPTION
**Pull request by olsonse**
I'm now back from datasheet land.  I think that this represents the datasheet much more cleanly and will avoid confusion much better.  These changes are somewhat aggressive--it will not generate exactly the same binary as before (some small changes in the event/channel/interrupt mapping were made).

The changes made to the examples rely on my changes in my pasm branch where -Idir (include dir command line option) and /\* */ comments were added.

Can you please look this over?  I have not yet had the time to run the examples yet and won't be able to for a day or so.

### CODE REVIEW

**Comment by fhunleth** [comment]
Cool. I'll add this to my todo list and get back to you.

**Comment by olsonse** [comment]
I tested this and everything seems to work.  So, the only thing you should have to do is to look at it for consistency and clarity.

**Comment by fhunleth** [comment]
I just tried your changes out and made the corresponding changes to my PRU application, and it's currently not working. I suspect that my problem is simple, but I'll need to look into it some more later. I like the changes that I've seen so far especially since the naming consistency has always bugged me. Have you thought about writing a short upgrade guide? I.e. code that looks like "XYZ" is now "ABC". I'm also a little concerned about the difference between the code and the documentation. The pdf documentation was getting stale before your change, but I think that this one safely breaks some text in it. I wonder if @jadonk has thoughts on documentation. It would be nice if it were a simple markdown document that we could keep uptodate.

**Comment by jadonk** [comment]
Can you put the documentation on the eLinux wiki? http://elinux.org/Ti_AM33XX_PRUSSv2

I think you should create a README.md (replacing README.txt) that contains out-of-box usage documentation and that it should refer to a page on the wiki for migration documentation. It should also make it clear that the code is based on a release from TI but is now maintained by the BeagleBoard.org community.

Side note: for the upcoming Debian image release, can you test the releases to make sure pasm works as you'd expect and provide a release tag with something that you've tested?

File: pru_sw/app_loader/interface/prussdrv.c [hunk 1]
```diff
@@ -415,10 +423,9 @@ int prussdrv_pruintc_init(const tpruss_intc_initdata *prussintc_init_data)
     pruintc_io[PRU_INTC_ESR2_REG >> 2] = mask2;
     pruintc_io[PRU_INTC_SECR2_REG >> 2] = mask2;
 
-    for (i = 0; i < MAX_HOSTS_SUPPORTED; i++)
-        if (prussintc_init_data->host_enable_bitmask & (1 << i)) {
-            pruintc_io[PRU_INTC_HIEISR_REG >> 2] = i;
-        }
+    for (i = 0; prussintc_init_data->hosts_enabled[i] < NUM_PRU_HOSTS; ++i)
```
Comment (fhunleth) [inline pru_sw/app_loader/interface/prussdrv.c]: It looks like the for loop should be the following based on the header file:

```
for (i = 0; i < NUM_PRU_HOSTS &&
     prussintc_init_data->hosts_enabled[i] != -1; ++i)
```

**Comment by fhunleth** [comment]
Amazingly my PRU code still doesn't run with these changes. I'm still looking for the bug. The problem is that poll() never returns for me. Normally this would indicate that the ARM is not receiving an interrupt from the PRU. I added quite a few error checks to the loader and double checked my application updates. Still nothing. Will debug more tomorrow.

**Comment by olsonse** [comment]
are you using all the changes on my events branch?  I had the same problem until I made the more recent changes.

**Comment by olsonse** [comment]
Also, with the for loop, I was relying on ((unsigned int) -1) > NUM_PRU_HOSTS.
hosts_enabled is an array of unsigned ints.

We could also write the loop as
for (i = 0; i < NUM_PRU_HOSTS &&
     prussintc_init_data->hosts_enabled[i] < NUM_PRU_HOSTS; ++i)

The comparison with NUM_PRU_HOSTS was to ensure that only input from the user that was in the correct range would get used--the first invalid entry (as -1 is) would cause the loop to stop.

**Comment by olsonse** [comment]
Incidentally, I was wondering which debugger you have been using.  I have been using prudebug quite successfully.

**Comment by olsonse** [comment]
@fhunleth and @jadonk :  Some thoughts for documentation...

First:
I've just committed a few notes that I wrote up about the changes we've made in the last while.  This should be a decent start at the content we need to change the documentation.

On a related note...
With regards to documentation, we have two types of changes:
1.  The changes to the prussdrv library.
2.  The few changes I made to PASM
   The documentation for the first set of changes and the rest of prussdrv does not currently belong in the reference guide.  With regards to the second, I am wondering if @jadonk were able to get the documentation changes incorporated with the reference guide since pasm is documented there.  _Or_ perhaps we ought to have a reference guide that does not include the pasm documentation, but just documents the hardware (the pasm documentation would then go wherever we put the rest of the documentation).  This might be the best path, especially since it is very likely that further changes to pasm could be made.

Pertaining to the documentation that does not belong in the reference guide, why don't we just host documentation with github pages.  My strong preference is that the community does not have multiple sites to go to for "official (community supported)" package _and_ documentation.  Also, I think the git model of contribution for api-docs works better for this than does a wiki.  Plus, github makes this very painless.
Advantages:
1.  URL would be simple: http://beagleboard.github.io/am335x_pru_package
2.  anyone can clone, change docs, then issue pull request
       --changes can be vetted or at least looked over for correctness
       --this is api documentation, not just wiki-style information
3.  github supports all sorts of markdown.
4.  Anyone who clones the package, also has the most recent, up-to-date, documentation.  Using github's markdown will really help in making things readable in text
5.  We don't have to worry about documentation spoilage (since git takes care of robustness of info already)
6.  We could use auto-generated api documentation (ala Doxygen).

Now that I think about this, I think it would be really good to take some of the info at http://elinux.org/Ti_AM33XX_PRUSSv2 and elsewhere for general information, _but_ use <b>Doxygen</b> for the actual API documentation.  I've already started adding Doxygen style comments in the code.  In my opinion, this is the absolute best way of documenting the code, since there is necessarily a single source of information (rather than both comments in the code as well as some other file for documentation).  **_Furthermore**_, the documentation could easily be turned into PDF or other various output formats using Doxygen.

**Comment by fhunleth** [comment]
@olsonse - I was testing using the 606d6f2 tag which had your latest, except for your changes from last night. Regarding the debugger, I haven't been using one, but it's good to hear that prudebug works well.

As for my thoughts on the docs, I'm in complete agreement about keeping the docs in git for the reasons you state. Doxygen API docs work for me. 

Also, sorry that I'm taking so long to test out your updates, but it's going to take me a little longer. I'm giving a talk soon on the code that uses the PRU, and I need to prep and gather more stats for that. 

To avoid holding up any progress, I generally like your changes despite them not working for me (which may or may not be something that you introduced). Regarding wording, I initially stumbled on the difference between "clearing" and "reseting" in the API. I had to go to the code to see what was meant so that I could call the right function. I have a minor dislike for name of the function prussdrv_pru_reset_event since it clears the event AND resets the associated interrupt. I'd probably delete it and make the user make two calls so that their code is more clear, but it's not a strong desire.

**Comment by jadonk** [comment]
My biggest concerns with making it too official is not supporting the upcoming C compiler (currently in closed beta--let me know if you want to be on it) and supporting the remote proc interface rather than UIO. If we have statements in the documentation that help address supporting those, I agree the page on github.io would look cleanest for hosting the documentation. If you have any quick-hints to help me set it up, let me know.

**Comment by fhunleth** [comment]
I'm not too worried about the C compiler, since the work has been in fixing up the loader. I assume that the C compiler can generate a binary similar to what pasm generates that we just load up into the PRU. I'm interested in hearing more about the availability of the remote proc interface. Are we wasting our time trying to clean up the UIO-based loader?

**Comment by jadonk** [comment]
@fhunleth I'm concerned we might be spending too much time on the UIO-based loader, yes. Perhaps it really is good enough and the work to clean-up the warts is good, but my feeling is that more likely we'll all just need to move the remote proc interface at some point. I'd prefer if you were working to get that interface better supported. 

Some links about the RemoteProc interface in some sort of chronological order:
- https://www.kernel.org/doc/Documentation/remoteproc.txt
- http://elinux.org/BeagleBone_6502_RemoteProc_cape
- http://elinux.org/images/a/ac/What's_Old_Is_New-_A_6502-based_Remote_Processor.pdf
- https://github.com/beagleboard/kernel/blob/3.8/patches/pru/0007-PRU-remote-proc-wip.patch
- https://github.com/pantoniou/testpru
- https://github.com/mranostay/ws28xx-lighting-pru
- http://ranostay.blogspot.com/2013/12/adventures-in-am335x-pru-lighting.html
- https://github.com/jadonk/pruduino

It really isn't clear to me what is going to land upstream as I don't see remote proc for PRU in here: https://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/tree/drivers/remoteproc

I _do_ see the UIO driver in here: https://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/tree/drivers/uio/uio_pruss.c

It is just that I've been repeatedly told that Remote Proc is the _right_ way and UIO is the _wrong_ way. Not sure if the TI kernel developers will ever push the Remote Proc stuff upstream.

Seems the only practical thing to do is support both. :-(

**Comment by fhunleth** [comment]
@jadonk Thanks for all of the links. I'd seen some of the remoteproc stuff before, and it was good to see a few more. I get why it's the _right_ way. To be honest, though, I'm feeling a little burnt following code that doesn't make it upstream (at work, it's video drivers, with the BBB it's little things like device tree overlays). I'm not sure that I personally want to invest time in it yet since it feels like a distraction from my main goals with the BBB. I'll suppose that I'll be dragged into remoteproc when the time is right.

As for this project, I like PASM and I really like @olsonse's cleanup to the UIO driver. I'm wondering if some focus would help things. What do you think about splitting the project so that we have a PASM repository, a UIO loader repository, and a remote proc one? The idea being that each project has a clear mandate so when we're working on UIO, we don't have to worry about remote proc, etc. I don't want to ignore the merits of remote proc, I just don't want to slow down the UIO work since it is supported upstream.

**Comment by olsonse** [comment]
I also like PASM and will not be using a c-compiler for my main needs when it comes out--I've fine-tuned the assmbler code I use to squeak out every last bit of timing control.  I am curious as to how the c-compiler work is being done.  Is it being done as a part of gcc where pasm essentially gets swallowed up into an architecture of the gnu assembler and gcc does most of the hard work already?  Or is it something much more on its own?

Also, I'm curious about what you both think about the remoteproc  interface.  My simple browsing shows a very simple API for remoteproc.  How do we do things like configuring the interrupt controller, signaling events, waiting on interrupts, memory mapping (i.e. read/writes), and so on?  These things are still necessary and we cannot simply rely on static memory writes (via ELF file sections).  So, unless I missed something, I don't see facilities for this in remoteproc directly--so would it not make more sense to just make remoteproc or UIO as the back-end of our loader library?

Also:  since @fhunleth's addition of the pru_exec_code function, I had the idea that it might actually be nice to have PASM provide a library interface where code could be compiled into binary on the fly.  That way, I could just keep the raw assembler inside my python code.  This would certainly be much less doable if we have to move to remoteproc and the program is an ELF file.

If I've missed something obvious, sorry for my lack of understanding remoteproc completely.

**Comment by jadonk** [comment]
Any more thoughts since I created http://beagleboard.github.io/am335x_pru_package ?

**Comment by fhunleth** [comment]
Hey all. Sorry for the delays - I've been travelling. I think that I'm just too busy right now to contribute much more to am335x_pru_package. At the moment, the PRU tools work for me and other aspects of my projects are more pressing. Hopefully I'll find a new project to justify hacking more on the PRU, since it is quite a bit of fun. I'll be watching the improvements to the package and hope to be back soon.

**Comment by jadonk** [comment]
@fhunleth No chance you'd mentor students in GSoC this summer then?

**Comment by fhunleth** [comment]
@jadonk Wow, mentoring students would be a huge honor. Thank you for the offer. My passion, though, has been advancing Erlang on the Beaglebone and other embedded devices via the nerves-project. I am currently at ErlangFactory and will be presenting a simple industrial camera that I built with the BBB, its PRUs, and a custom camera cape tomorrow. I will think more about your offer in case there's a way for it to mutually benefit both projects. Also, I just spoke with Francesco Cesarini from Erlang Solutions who was telling me that there would be support coming to the new Beaglebone releases, and I was curious about that. Feel free to continue the conversation off github, since I'm not sure how many people we're spamming now.

**Comment by fhunleth** [comment]
Closing since this is 4 years old and I'm not sure anyone uses this repo any more.

**Pull request closed without merging**